### PR TITLE
fix(web): Saved views UX improvements

### PR DESCRIPTION
**Pull request by bezbac**
**Disclaimer**: Experimental PR review
---

<h3>Greptile Summary</h3>

This PR improves the UX of saved views by ensuring that filter facets remain visible and marked as active even before the backend has returned the matching option values — covering categorical facets, key-value facets, and the auto-expansion of relevant filter accordions on view load.

- **`CategoricalFacet`** now merges currently-selected `value` items into `visibleOptionValues` so that a selection from a saved view is always rendered in the checkbox list, not silently hidden when the backend returns an empty options list.
- **`KeyValueFilterBuilder`** gains a `lodash/isEqual`-gated `useEffect` to sync `localFilters` when `activeFilters` changes externally, and merges selected keys from local state into the combobox options so previously-chosen keys remain visible.
- **`useTableViewManager`** auto-expands filter accordion items that correspond to columns in an applied view's filters, and all table use-cases are updated to pass `setExpandedFilters` / `expandableFilterColumns` / `currentExpandedFilters` as well as to use `explicitFilterState` over the broader `filterState`.

<h3>Confidence Score: 4/5</h3>

Safe to merge; the changes are well-scoped UX improvements backed by new tests, with one minor isActive edge case worth discussing

The core logic — surfacing selected values during loading and auto-expanding filter facets from saved views — is sound and covered by new client tests. The only concern is in useSidebarFilterState: the terminal standalone hasExplicitCheckboxFilter condition lacks an operator guard, so a facet where the user has selected every available option under 'any of' now shows as active (with a Clear button) whereas it previously did not. It won't corrupt data or break navigation, but may confuse users who see a Clear badge on a facet that effectively filters nothing.

web/src/features/filters/hooks/useSidebarFilterState.tsx — specifically the new isActive expression around line 1600

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| web/src/features/filters/hooks/useSidebarFilterState.tsx | Adds helpers to merge active filter values into available options for key-value facets and fixes isActive for loading states; the standalone hasExplicitCheckboxFilter condition slightly over-broadens the active check for "any of" + all-values-selected case |
| web/src/components/table/data-table-controls.tsx | CategoricalFacet now includes currently-selected values in the visible options list even when the backend returns none, correctly fixing the empty-options UX issue |
| web/src/components/table/key-value-filter-builder.tsx | Adds a deep-equality-gated useEffect to sync localFilters with external activeFilters changes; merges selected keys from local state into the combobox options |
| web/src/components/table/table-view-presets/hooks/useTableViewManager.ts | Adds setExpandedFilters + currentExpandedFilters wiring; when a saved view is applied, filter facets matching expandableFilterColumns are auto-expanded |
| web/src/components/table/data-table-controls.clienttest.tsx | New client tests validating that CategoricalFacet renders selected values and the Clear button when the backend returns no options |
| web/src/components/table/key-value-filter-builder.clienttest.tsx | New client tests covering external activeFilters sync, missing key display in combobox, and draft preservation across equivalent rerenders |
| web/src/components/table/use-cases/traces.tsx | Wires setExpandedFilters, expandableFilterColumns, and currentExpandedFilters into useTableViewManager for the traces table |
| web/src/components/table/use-cases/observations.tsx | Same expanded-filter wiring as traces for the observations table; switches from filterState to explicitFilterState |
| web/src/components/table/use-cases/sessions.tsx | Adds expanded-filter wiring and switches to explicitFilterState for the sessions table |
| web/src/components/table/use-cases/scores.tsx | Adds expanded-filter wiring and switches to explicitFilterState for the scores table |
| web/src/features/events/components/EventsTable.tsx | Adds expanded-filter wiring and switches to explicitFilterState for the events table |
| web/src/features/experiments/components/table/ExperimentsTable.tsx | Switches from filterState to explicitFilterState and adds expanded-filter wiring consistent with the other table use-cases |
| web/src/features/experiments/components/table/ExperimentItemsTable.tsx | Same explicitFilterState switch and expanded-filter wiring for experiment items table |
| web/src/components/session/index.tsx | Wires setExpandedFilters and expandableFilterColumns into useTableViewManager for the session events page; switches to explicitFilterState |

</details>

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[User applies Saved View] --> B[useTableViewManager.applyViewState]
    B --> C[Validate filters]
    C --> D{expandableFilterColumns configured?}
    D -- Yes --> E[Compute nextExpandedFilters from validFilters columns]
    E --> F[setExpandedFilters]
    F --> G[useSidebarFilterState expandedSet updated]
    D -- No --> H[setFilters]
    G --> H
    H --> I[useSidebarFilterState recomputes UIFilters]
    I --> J{Backend options loaded?}
    J -- Yes --> K[selectedValues from options and filter.value]
    J -- No --> L[CategoricalFacet shows value from visibleOptionValues]
    J -- No --> M[hasExplicitCheckboxFilter marks facet isActive]
    K --> N[Normal render with counts]
    L --> N
    M --> O[Clear badge shown even during loading]
```

<details open><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `web/src/features/filters/hooks/useSidebarFilterState.tsx`, line 1600-1609 ([link](https://github.com/langfuse/langfuse/blob/3ec8f2c84a130086bf3edb29a7018c62d2ff8a48/web/src/features/filters/hooks/useSidebarFilterState.tsx#L1600-L1609)) 

   <a href="#">P2</a> **Standalone `hasExplicitCheckboxFilter` makes all-selected "any of" facets appear active**

   The terminal `hasExplicitCheckboxFilter` guard has no operator constraint, so it fires when `checkboxFilter` exists with values regardless of operator. This changes a pre-existing intentional behaviour: when a user selects **all** available options under the `"any of"` operator, the old code set `isActive = false` (semantically equivalent to no filter, so no Clear badge). Now `hasExplicitCheckboxFilter` is `true` in that case, so the facet shows as active and renders a Clear button that does nothing semantically useful. The "none of" loading-state fix should be scoped to the actual problem by only applying `hasExplicitCheckboxFilter` when `selectedValues` is empty (i.e., during loading before the backend has returned any options).

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: web/src/features/filters/hooks/useSidebarFilterState.tsx
   Line: 1600-1609

   Comment:
   **Standalone `hasExplicitCheckboxFilter` makes all-selected "any of" facets appear active**

   The terminal `hasExplicitCheckboxFilter` guard has no operator constraint, so it fires when `checkboxFilter` exists with values regardless of operator. This changes a pre-existing intentional behaviour: when a user selects **all** available options under the `"any of"` operator, the old code set `isActive = false` (semantically equivalent to no filter, so no Clear badge). Now `hasExplicitCheckboxFilter` is `true` in that case, so the facet shows as active and renders a Clear button that does nothing semantically useful. The "none of" loading-state fix should be scoped to the actual problem by only applying `hasExplicitCheckboxFilter` when `selectedValues` is empty (i.e., during loading before the backend has returned any options).

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 1 code review issue. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 1
web/src/features/filters/hooks/useSidebarFilterState.tsx:1600-1609
**Standalone `hasExplicitCheckboxFilter` makes all-selected "any of" facets appear active**

The terminal `hasExplicitCheckboxFilter` guard has no operator constraint, so it fires when `checkboxFilter` exists with values regardless of operator. This changes a pre-existing intentional behaviour: when a user selects **all** available options under the `"any of"` operator, the old code set `isActive = false` (semantically equivalent to no filter, so no Clear badge). Now `hasExplicitCheckboxFilter` is `true` in that case, so the facet shows as active and renders a Clear button that does nothing semantically useful. The "none of" loading-state fix should be scoped to the actual problem by only applying `hasExplicitCheckboxFilter` when `selectedValues` is empty (i.e., during loading before the backend has returned any options).


`````

</details>

<sub>Reviews (1): Last reviewed commit: ["fix(web): Always consider saved view fil..."](https://github.com/langfuse/langfuse/commit/3ec8f2c84a130086bf3edb29a7018c62d2ff8a48) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=30801586)</sub>

### CODE REVIEW

**Comment by bezbac** [comment]
@greptileai

**Comment by bezbac** [comment]
@claude review

File: web/src/components/table/data-table-controls.tsx [hunk 1]
```diff
@@ -597,19 +597,22 @@ export function CategoricalFacet({
...
         return (
           option.toLowerCase().includes(search) ||
           displayLabel.toLowerCase().includes(search)
         );
       })
-    : options;
+    : visibleOptionValues;
 
   const hasMoreFilteredOptions = filteredOptions.length > MAX_VISIBLE_OPTIONS;
   const visibleOptions = showAll
```
[the remark this answers is not in the document]
Comment (bezbac) [reply web/src/components/table/data-table-controls.tsx:618]: This is okay as `hasMoreOptions` would be true.

**Review by nimarb — APPROVED** [review]
